src/FileStorageTools/Adapter/LocalAdapter.php

```diff
@@ -0,0 +1,138 @@
+use BayWaReLusy\FileStorageTools\Exception\DirectoryNotEmptyException;
+use BayWaReLusy\FileStorageTools\Exception\FileCouldNotBeOpenedException;
+use BayWaReLusy\FileStorageTools\Exception\LocalFileNotFoundException;
+use BayWaReLusy\FileStorageTools\Exception\ParentNotFoundException;
+use BayWaReLusy\FileStorageTools\Exception\RemoteFileDoesntExistException;
+use BayWaReLusy\FileStorageTools\Exception\UnknownErrorException;
+
+class LocalAdapter implements FileStorageAdapterInterface
+{
+    public function __construct(
+        protected string $remotePath
+    ) {
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function createDirectory(string $path): void
+    {
+        try {
+            if (!file_exists($this->remotePath)) {
+                mkdir($this->remotePath, 0777, true);
+            }
+            if (file_exists($this->remotePath . $path)) {
+                throw new DirectoryAlreadyExistsException("The directory already exists");
+            }
+            if (!@mkdir($this->remotePath . $path, 0777, false)) {
+                throw new ParentNotFoundException("The parent directory could not be found");
+            }
+        } catch (ParentNotFoundException | DirectoryAlreadyExistsException $e) {
+            throw $e;
+        } catch (\Throwable $e) {
+            throw new UnknownErrorException("Unexpected error");
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function deleteDirectory(string $path): void
+    {
+        try {
+            if (!file_exists($this->remotePath . $path)) {
+                throw new DirectoryDoesntExistsException("the directory doesn't exists");
+            }
+            if (!@rmdir($this->remotePath . $path)) {
+                throw new DirectoryNotEmptyException("The directory isn't empty");
+            }
+        } catch (DirectoryDoesntExistsException | DirectoryNotEmptyException $e) {
+            throw $e;
+        } catch (\Throwable $e) {
+            throw new UnknownErrorException("Unknown exception");
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function uploadFile(string $directory, string $pathToFile): void
+    {
+        //Make a directory our "remote" place
+        if (!file_exists($this->remotePath)) {
+            $this->createDirectory($this->remotePath);
+        }
+        // Check first if local file exists and can be opened
+        if (!file_exists($pathToFile)) {
+            throw new LocalFileNotFoundException("File not found.");
+        }
+        if (!$fileContent = file_get_contents($pathToFile)) {
+            throw new FileCouldNotBeOpenedException("File couldn't be read.");
+        }
+        if (!file_exists($this->remotePath . $directory)) {
+            throw new ParentNotFoundException("Remote parent could not be found");
+        }
+        try {
+            file_put_contents($this->remotePath . $directory . '/' . basename($pathToFile), $fileContent);
```

Comment:
Two problems here:

1. `file_get_contents()` and `file_put_contents()` requires the file content to be put in memory. This is probably much slower than using `fopen()` & `copy()`.
2. `file_put_contents()` returns `int|false`. So this needs to be checked, because this function probably doesn't throw any exceptions. Return needs to be checked very carefully for this function, because "0" means success, but "false" doesn't.